src/interpreter/balances.rs

```diff
@@ -0,0 +1,191 @@
+}
+
+impl From<&CheckedTransaction> for RuntimeBalances {
+    fn from(tx: &CheckedTransaction) -> Self {
+        let iter = tx.free_balances().map(|(asset, value)| (*asset, *value));
+
+        Self::try_from_iter(iter).expect(
+r#"This is a bug!
+
+A checked transaction shouldn't produce a malformed initial free balances set.
+
+Please, file a report mentioning an incorrect transaction validation implementation that allowed a type-safe checked transaction to be created from a malformed inputs set."#)
+    }
+}
+
+impl RuntimeBalances {
+    /// Attempt to create a set of runtime balances from an iterator of pairs.
+    ///
+    /// This will fail if, and only if, the provided asset/balance pair isn't consistent or a
+    /// balance overflows.
+    pub fn try_from_iter<T>(iter: T) -> Result<Self, ValidationError>
+    where
+        T: IntoIterator<Item = (AssetId, Word)>,
+    {
+        iter.into_iter()
+            .sorted_by_key(|k| k.0)
+            .enumerate()
+            .try_fold(HashMap::new(), |mut state, (i, (asset, balance))| {
+                let offset = VM_MEMORY_BALANCES_OFFSET + i * (AssetId::LEN + WORD_SIZE);
+
+                state
+                    .entry(asset)
+                    .or_insert(Balance::new(0, offset))
+                    .checked_add(balance)
+                    .ok_or(ValidationError::ArithmeticOverflow)?;
+
+                Ok(state)
+            })
+            .map(|state| Self { state })
+    }
+
+    /// Fetch the balance of a given Id, if set.
+    pub fn balance(&self, asset: &AssetId) -> Option<Word> {
+        self.state.get(asset).map(Balance::value)
+    }
+
+    fn _set_memory_balance(balance: &Balance, memory: &mut [u8]) -> Word {
+        let value = balance.value();
+        let offset = balance.offset();
+
+        let offset = offset + AssetId::LEN;
+        let memory = &mut memory[offset..offset + WORD_SIZE];
+
+        memory.copy_from_slice(&value.to_be_bytes());
+
+        value
+    }
+
+    /// Attempt to add the balance of an asset, updating the VM memory in the appropriate
+    /// offset
+    pub fn checked_balance_add(&mut self, memory: &mut [u8], asset: &AssetId, value: Word) -> Option<Word> {
+        self.state
+            .get_mut(asset)
+            .and_then(|b| b.checked_sub(value))
+            .map(|balance| Self::_set_memory_balance(balance, memory))
+            .or((value == 0).then(|| 0))
+    }
+
+    /// Attempt to subtract the balance of an asset, updating the VM memory in the appropriate
+    /// offset
+    pub fn checked_balance_sub(&mut self, memory: &mut [u8], asset: &AssetId, value: Word) -> Option<Word> {
+        self.state
+            .get_mut(asset)
+            .and_then(|b| b.checked_sub(value))
+            .map(|balance| Self::_set_memory_balance(balance, memory))
+            .or((value == 0).then(|| 0))
+    }
+
+    /// Write all assets into the VM memory.
+    pub fn to_vm<S>(self, vm: &mut Interpreter<S>) {
```

Comment:
This seems like a very impure and leaky encapsulation of logic, as the correctness is heavily reliant on the external / global state of the VM and the order of operations when initializing the stack outside of this call as well as manipulating REG_SP directly.

Instead, I think it would be cleaner to pass a mutable slice to write the balances state into, rather than a mutable reference to the whole VM. That way, the logic for setting reserving the stack and managing REG_SP is consolidated into one place rather than delegated away.

Reply:
We don't have a use-case where we update a slice of bytes with the balances - except when this slice of bytes happens to be the memory representation of the VM. This means we don't really update slices, but instead, update the VM itself. Keep the function like this unmistakably translates into that - while receiving slices allows multiple incorrect interpretations of what this function is supposed to do

Incrementing the stack pointer here is also an exception since we don't really do that anywhere else but during assets initialization. Also, its worthy to keep in mind that this balances struct definition is part of the interpreter module - therefore, it is ok to access the VM internals.

Reply:
> this balances struct definition is part of the interpreter module - therefore, it is ok to access the interpreter internals

In that case, it seems like it may be a good idea to make this struct `pub(crate)` instead of `pub` so that it doesn't get misused outside the intended context. But It's not worth blocking the PR on this point.